design-documents/csp.md

```diff
@@ -0,0 +1,169 @@
+## Content Security Policies
+CSP allows a server to define a list of trusted sources of static files (scripts, styles, fonts, media files etc.)
+along with disabling certain unsafe HTML features like inline JavaScript. It's a powerful tool
+in the fight with XSS attacks.
+## Modes
+CSP supports two modes:
+* restrict mode `Content-Security-Policy`
+* report violations mode `Content-Security-Policy-Report-Only`
+ 
+1st one would follow defined rules and prevent certain scenarios from happening
+(like not loading a font from an untrusted address).
+ 
+Second one will only report violations
+to the server but proceed as it would without these restrictions. This mode would be convenient to merchants
+just adopting CSP on their existing stores to see what kind of legit resources/features they need to enable
+via CSP.
+## Backward compatibility
+In order to remind merchants that having CSP defined is important it will be enabled by default. But since adopting it
+would require manually whitelisting trusted resource/features by merchants and extension developers CSP
+will be functioning in the report-only mode. This will allow us to introduce CSP in a patch release.
+ 
+## API
+At the beginning Magento would not manage all of the possible CSP policies, that's
+why we need an extensible SPI to allow 3rd party developers to implement and control new
+policies.
+ 
+CSP API/SPI will consist of the following interfaces:
+ 
+* Policy DTO. The interface has only policy ID because not all CSP policies look like _\<policy name\> \<value\> \<value\>_,
```

Comment:
Could you provide some details about different policy formats? I did not find any information about it in RFC. Or are you talking about serialized SCP?

Reply:
examples would include `plugin-types`, `report-to`, `sandbox` and some other policies that either require additional headers to work or at least accept different values than URIs and preserved 'self'/'unsafe-inline' words like *-src policies do

Reply:
Probably I'm not clear understand but `plugin-types`, `report-to`, `sandbox` have the same format `<policy name/> <value/> <value/>`

Reply:
I got confused, though the question was related to the limitation of available policies through csp_whitelist.xml - for that file allowed options for values are important as well. `report-to` requires additional header and Subresource Integrity may involve `Access-Control-Allow-Origin` header to be implemented